types/prefs/struct_list.go

```diff
@@ -0,0 +1,180 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+package prefs
+
+import (
+	"reflect"
+	"slices"
+
+	jsonv2 "github.com/go-json-experiment/json"
+	"github.com/go-json-experiment/json/jsontext"
+	"tailscale.com/types/opt"
+	"tailscale.com/types/ptr"
+	"tailscale.com/types/views"
+)
+
+// StructList is a preference type that holds zero or more potentially mutable struct values.
+type StructList[T views.Cloner[T]] struct {
+	preference[[]T]
+}
```

Comment:
curious - can this be folded into the existing `List` type? or are the nested types strictly incompatible?

Reply:
Yeah, it is very unfortunate that it is not possible to fold the `StructList` and `StructMap` types with the `List` and `Map`, as we have different underlying view types for slices/maps of mutable (and therefore cloneable) and non-mutable types.

Please see [`[views.Slice]`](https://github.com/tailscale/tailscale/blob/1e8f8ee5f160478015b374aead80ee6aac789577/types/views/views.go#L203) vs [`views.SliceView`](https://github.com/tailscale/tailscale/blob/1e8f8ee5f160478015b374aead80ee6aac789577/types/views/views.go#L141) and [`[views.Map]`](https://github.com/tailscale/tailscale/blob/1e8f8ee5f160478015b374aead80ee6aac789577/types/views/views.go#L422) vs [`[views.MapFn]`](https://github.com/tailscale/tailscale/blob/1e8f8ee5f160478015b374aead80ee6aac789577/types/views/views.go#L516).